lib/addressable/idna/pure.rb

```diff
@@ -97,7 +97,7 @@ def self.to_unicode(input)
         if part =~ /^#{ACE_PREFIX}(.+)/
           begin
             punycode_decode(part[/^#{ACE_PREFIX}(.+)/, 1])
-          rescue Addressable::IDNA::PunycodeBadInput
+          rescue Addressable::IDNA::Pure::PunycodeBadInput
```

Comment:
Oh, it is possible to not change the namespace of these errors? If we are not going for a major version now. I know at least one library (authored by me in the past, hehe) that rely on these names (well, [only one so far](https://github.com/twingly/twingly-url/issues/112)): https://github.com/twingly/twingly-url/blob/e20f5fce077d93e89ef8520961be453c90cfec8c/lib/twingly/url.rb#L21

Reply:
I guess having these exceptions defined under `Addressable::IDNA::Pure` is more correct though, if they can only be raised from code in `Addressable::IDNA::Pure`? 

Can `Addressable::IDNA::Libidn2` raise any exceptions? Looks like `to_ascii` can raise `RuntimeError`. `to_unicode` being discussed at https://github.com/sporkmonger/addressable/pull/496#discussion_r1145256087

Reply:
Good point, It could make sense to move these exception back up in `IDNA` and if possible re-use them in `libidn1` and `libidn2` then. Though at the moment I made a very simple error handling so only the message changes. Distinguishing between different types of errors in `libidn2` would require keeping a mapping of native error code ([example from ruby-idna](https://github.com/ogom/ruby-idna/blob/master/lib/idna/error.rb)), I'd rather avoid that if possible.

Here is what I suggest: let's add a generic `Addressable::IDNA::Error` (which I will use in `libidn2` instead of `RuntimeError`, it's actually something I wanted to do but forgot) and move back the pure exceptions to IDNA to keep compatibility but make them inherit from `Addressable::IDNA::Error`. This way the old exception don't change, and people can rescue `Addressable::IDNA::Error` if they want all of them (this one is future-proof).

I just did that in 84fdb89ef413f60b219f89c9910604247d3b84f0